### PR TITLE
[BXMSPROD-1770] Added OS preparation step

### PR DESCRIPTION
**Pull request by lampajr**
JIRA:
- https://issues.redhat.com/browse/BXMSPROD-1770

referenced Pull Requests:
- https://github.com/kiegroup/kogito-pipelines/pull/590

Added OS preparation step.

If we agree on this solution I'll revert the commit [9346bcd](https://github.com/kiegroup/optaplanner/pull/2146/commits/9346bcd1616013541c5f8ed4613030ced630c875) and propagate the same fix on all other kogito-related projects.

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] tests</b>
  
- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: 
    please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: 
    add the label `run_fdb`

- for a <b>compile downstream build</b>  
  please add comment: <b>Jenkins run cdb</b>

- for a <b>full production downstream build</b>  
  please add comment: <b>Jenkins execute product fdb</b>

- for an <b>upstream build</b>  
  please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] mandrel</b>
</details>

### CI Status

 You can check OptaPlanner repositories CI status from [Chain Status webpage](https://kiegroup.github.io/optaplanner/).

### CODE REVIEW

**Comment by lampajr** [comment]
Before **merging** commit 9346bcd1616013541c5f8ed4613030ced630c875 must be reverted.

**Comment by lampajr** [comment]
jenkins retest this

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -31,13 +31,22 @@ defaults:
...
 
 jobs:
+  os-prep:
+    runs-on: ubuntu-latest
+    outputs:
+      os: ${{ steps.set-os.outputs.operating-systems }}
+    steps:
+      - name: OS Preparation
+        id: set-os
+        uses: lampajr/kogito-pipelines/.ci/actions/os-preparation@BXMSPROD-1770_os_preparation
```
Comment (mareknovotny) [inline .github/workflows/pull_request.yml:41]: here don't forget to use kiegroup instead of your github

Comment (lampajr) [reply .github/workflows/pull_request.yml:41]: yeah sure, once the https://github.com/kiegroup/kogito-pipelines/pull/590 will be merged in `kogito-pipelines` I'll fix here checking everything will be fine with the correct configuration.

**Review by radtriste — APPROVED** [review]

File: .github/workflows/pull_request.yml [hunk 2]
```diff
@@ -31,13 +31,22 @@ defaults:
...
+        uses: kiegroup/kogito-pipelines/.ci/actions/os-preparation@main
   build-chain:
+    needs: os-prep
     concurrency:
       group: pull_request-${{ github.event_name }}-${{ github.head_ref }}-${{ matrix.os }}-${{ matrix.java-version }}
       cancel-in-progress: true
     strategy:
       matrix:
-        os: [ ubuntu-latest, windows-latest ]
+        os: ${{ fromJSON(needs.os-prep.outputs.os) }}
```
Comment (triceo) [inline .github/workflows/pull_request.yml:49]: Why would we do it like this? We always want to run on Windows. What's the reason not to?

**Review by triceo — COMMENTED** [review]
I'm not sure what we're trying to do here.

File: .github/workflows/pull_request.yml (hunk 2 above)
Comment (radtriste) [reply .github/workflows/pull_request.yml:49]: It is using file from https://github.com/kiegroup/kogito-pipelines/blob/main/.github/matrix-os.json to determine the os to run on (so still ubuntu and windows)

Comment (triceo) [reply .github/workflows/pull_request.yml:49]: Then what's the `windows_check` label for?

Comment (radtriste) [reply .github/workflows/pull_request.yml:49]: will be handled in https://github.com/kiegroup/kogito-pipelines/blob/main/.github/matrix-os.json#L12

Comment (lampajr) [reply .github/workflows/pull_request.yml:49]: Hello @triceo , the idea behind this was to centralize the Operating Systems configuration for kogito projects in a single place, i.e., [https://github.com/kiegroup/kogito-pipelines/](https://github.com/kiegroup/kogito-pipelines/). For all those projects that use by default both windows and ubuntu won't need to add the `windows_check` label, but all the others yes (if needed). Another thing is that by centralizing this configuration, whenever there will be the need to update the OS config for pull requests, we won't have to do it on every project but only here https://github.com/kiegroup/kogito-pipelines/blob/main/.github/matrix-os.json

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -52,6 +52,9 @@ How to retest this PR or trigger a specific build:
   - for <b>github actions</b> job: 
     add the label `run_fdb`
 
+- for a <b>windows build</b>
+  please add the label `windows_check`
```
Comment (triceo) [inline .github/pull_request_template.md:56]: May I suggest we remove this, as OptaPlanner will not be using this? 

```suggestion
- for a <b>windows build</b>
  please add the label `windows_check`
```

Comment (lampajr) [reply .github/pull_request_template.md:56]: Looks ok to me

**Pull request merged**